src/vs/platform/contextkey/common/scanner.ts

```diff
@@ -0,0 +1,387 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { CharCode } from 'vs/base/common/charCode';
+import { illegalArgument } from 'vs/base/common/errors';
+
+export enum TokenType {
+	LParen = '(',
+	RParen = ')',
+
+	Neg = '!',
+
+	Eq = '==',
+	NotEq = '!=',
+
+	Lt = '<',
+	LtEq = '<=',
+	Gt = '>',
+	GtEq = '>=',
+
+	RegexOp = '=~',
+
+	RegexStr = 'RegexStr',
+
+	True = 'true',
+	False = 'false',
+
+	In = 'in',
+	Not = 'not',
+
+	And = '&&',
+	Or = '||',
+
+	Str = 'Str',
+
+	QuotedStr = 'QuotedStr',
+
+	Error = 'ErrorToken',
+
+	EOF = 'EOF'
+}
+
+export type Token = {
```

Comment:
Once you switch to a const enum, you can write this as an or type (e.g. `type Token = { type: TokenType.Str; offset: number; lexeme: string} | ... ) and add strong typing around which tokens have a `lexeme` or not. This would allow removing a lot of `!` from the code.